app/database.js

```diff
@@ -9,72 +9,206 @@
+				}
+			}
+			callback();
+		});
+	}
+
+	function synchronizeDatasets(userId, callback) {
+			synchronizing = true;
+			var localdatasets = database.getQuery('getUserDatasets',[userId]);
+			var remotetolocaldatasets = [];
+			database.getOnlineQuery('getUserDatasets',[userId], function(remotedatasets) {
+				for (var i=0; i<localdatasets.length;i++) {
+					if (!localdatasets[i].dataset_online) {
+						lastId = localdatasets[i].dataset_id;
+					}
+				}
+				if (lastId==0) {
+					callback();
+				}
+
+				// Compare local with online
+				for (var i=0; i< localdatasets.length; i++) {
+					if (!localdatasets[i].dataset_online) {
+						pushDatasetOnline(localdatasets[i], callback);
+					} else {
+						var remote = $.grep(remotedatasets, function(e) { return e.dataset_id === localdatasets[i].dataset_id; });
+						if (remote.length !== 0) {
+							synchronizeDataset(localdatasets[i], remote);
+						}
+					}
+				}
+				// Compare online with local
+				for (var j=0 ;j<remotedatasets.length; j++){
+					var remote = $.grep(localdatasets, function(e) { return e.dataset_id === remotedatasets[j].dataset_id; });
+					if (remote.length === 0) {
+						pushDatasetLocal(remotedatasets[j]);
+					}
+				}
+				database.close();
+			});
+	}
+
+	function pushDatasetOnline(dataset, callback) {
+		var local_id = dataset.dataset_id;
+		dataset.dataset_online = 1;
+		var dataset = $.map(dataset, function(val, key) { if (key!="dataset_id") { return val; } });
+		database.executeQuery('addDataset', dataset, false, true);
+		database.lastInsertIdOnline('tbldatasets', 'dataset_id', function (id) {
+			database.executeQuery('updateDatasetId', [id, local_id], true, false);
+			if (local_id==lastId) {
+				callback();
+			}
+		});
+	}
+
+	function pushDatasetLocal(dataset) {
+		var dataset = $.map(dataset, function(val, key) { return (key=="dataset_date" || key=="dataset_lastedited") ? date.formatDatetime(val) : val; });
+		database.executeQuery('addDatasetAll', dataset, true, false);
+	}
+
+	function synchronizeDataset(local,remote){
+		var localTime = Date.parse(local.dataset_lastedited);
+		var onlineTime = Date.parse(remote[0].dataset_lastedited);
+		var recent = localTime - onlineTime;
+		if (recent > 0) {
+			local = $.map(local, function(val, key) { return (key=="dataset_date" || key=="dataset_lastedited") ? date.formatDatetime(val) : val; });
+			database.executeQuery('deleteDatasetbyId',[remote[0].dataset_id], false, true);
+			database.executeQuery('addDatasetAll', local, false, true);
+		} else if (recent < 0) {
+			remote = $.map(remote[0], function(val, key) { return (key=="dataset_date" || key=="dataset_lastedited") ? date.formatDatetime(val) : val; });
+			database.executeQuery('deleteDatasetbyId',[local.dataset_id], true, false);
+			database.executeQuery('addDatasetAll', remote, true, false);
+		}
+	}
+
+	function initOnlineDB(){
+		if (database.online()) {
+			db_online = mysql.createConnection({
```

Comment:
If you often need all the parameters in this same dictionary you might want to consider just returning a database_configuration dictionary form the configuration file. Or if this is the only place you require those parameters.